src/bytes.ts

```diff
@@ -0,0 +1,264 @@
+    const n1 = c1 - (c1 < 58 ? 48 : 87);
+    const n2 = c2 - (c2 < 58 ? 48 : 87);
+
+    bytes[i] = n1 * 16 + n2;
+  }
+
+  return bytes;
+}
+
+/**
+ * Convert a `bigint` to a `Uint8Array`.
+ *
+ * @param value - The bigint to convert to bytes.
+ * @returns The bytes as `Uint8Array`.
+ */
+export function bigIntToBytes(value: bigint): Uint8Array {
+  assert(typeof value === 'bigint', 'Value must be a bigint.');
+
+  const hex = value.toString(16);
+  return hexToBytes(hex);
+}
+
+/**
+ * Convert a `number` to a `Uint8Array`.
+ *
+ * @param value - The number to convert to bytes.
+ * @returns The bytes as `Uint8Array`.
+ * @throws If the number is not a safe integer.
+ */
+export function numberToBytes(value: number): Uint8Array {
+  assert(typeof value === 'number', 'Value must be a number.');
+  assert(
+    Number.isSafeInteger(value),
+    'Value is not a safe integer. Use `bigIntToBytes` instead.',
+  );
+
+  const hex = value.toString(16);
+  return hexToBytes(hex);
+}
+
+/**
+ * Convert a `string` to a UTF-8 encoded `Uint8Array`.
+ *
+ * @param value - The string to convert to bytes.
+ * @returns The bytes as `Uint8Array`.
+ */
+export function stringToBytes(value: string): Uint8Array {
+  assert(typeof value === 'string', 'Value must be a string.');
+
+  return new TextEncoder().encode(value);
+}
+
+/**
+ * Convert a byte-like value to a `Uint8Array`. The value can be a `Uint8Array`,
+ * a `bigint`, a `number`, or a `string`.
+ *
+ * If the value is a `string`, and it is prefixed with `0x`, it will be
+ * interpreted as a hexadecimal string. Otherwise, it will be interpreted as a
+ * UTF-8 string. To convert a hexadecimal string to bytes without interpreting
+ * it as a UTF-8 string, use {@link hexToBytes} instead.
+ *
+ * If the value is a `Uint8Array`, it will be returned as-is.
+ *
+ * @param value - The value to convert to bytes.
+ * @returns The bytes as `Uint8Array`.
+ */
+export function valueToBytes(value: Bytes): Uint8Array {
+  if (typeof value === 'bigint') {
+    return bigIntToBytes(value);
+  }
+
+  if (typeof value === 'number') {
+    return numberToBytes(value);
+  }
+
+  if (typeof value === 'string') {
+    if (value.startsWith('0x')) {
+      return hexToBytes(value);
+    }
+
+    return stringToBytes(value);
+  }
+
+  if (isBytes(value)) {
+    return value;
+  }
+
+  throw new Error(`Unsupported value type: "${typeof value}".`);
+}
+
+/**
+ * Concatenate multiple byte-like values into a single `Uint8Array`. The values
+ * can be `Uint8Array`, `bigint`, `number`, or `string`. This uses
+ * {@link valueToBytes} under the hood to convert each value to bytes. Refer to
+ * the documentation of that function for more information.
+ *
+ * @param values - The values to concatenate.
+ * @returns The concatenated bytes as `Uint8Array`.
+ */
+export function concatBytes(values: Bytes[]): Uint8Array {
+  return values.map(valueToBytes).reduce((previousValue, currentValue) => {
```

Comment:
You're allocating O(n) arrays in reducing, copying all elements from the previous one on each step. That's O(n^2) algorithm. This brings it down to O(n)

```typescript
let totalLength = 0;

const converted = new Array(values.length);
for (let i = 0; i < values.length; i++) {
  converted[i] = valueToBytes(values[i]);
  totalLength += converted[i].length;
}
// Not sure if Uint8Array has concat, assuming not:
const bytes = new Uint8Array(totalLength);
let at = 0;
let i = 0;
while (i < values.length) {
  bytes.set(converted[i], at);
  at += converted[i].length;
  i++;
}
```